packages/java-parser/src/comments.js

```diff
@@ -0,0 +1,156 @@
+  const headComments = [];
+  while (
+    attachComments.length > 0 &&
+    attachComments[0].endOffset < firstToken.startOffset
+  ) {
+    headComments.push(attachComments[0]);
+    attachComments.splice(0, 1);
+  }
+
+  if (headComments.length > 0) {
+    firstToken.leadingComments = headComments;
+  }
+
+  // edge case: when the file end with comments, it attaches as trailingComments to the last token
+  const lastToken = tokens[tokens.length - 1];
+  const tailComments = [];
+  while (
+    attachComments.length > 0 &&
+    attachComments[attachComments.length - 1].startOffset > lastToken.endOffset
+  ) {
+    tailComments.push(attachComments[attachComments.length - 1]);
+    attachComments.splice(attachComments.length - 1, 1);
+  }
+
+  if (tailComments.length > 0) {
+    lastToken.trailingComments = tailComments.reverse();
+  }
+
+  let currentToken = 0;
+  attachComments.forEach(element => {
+    // find the correct position to place the comment
+    while (
+      !(
+        element.startOffset > tokens[currentToken].endOffset &&
+        element.endOffset < tokens[currentToken + 1].startOffset
+      )
+    ) {
+      currentToken++;
+    }
+
+    // attach comment to the closest token
+    if (
+      element.startOffset - tokens[currentToken].endOffset <
+      tokens[currentToken + 1].startOffset - element.endOffset
+    ) {
+      if (!tokens[currentToken].hasOwnProperty("trailingComments")) {
+        tokens[currentToken].trailingComments = [];
+      }
+      tokens[currentToken].trailingComments.push(element);
+    } else {
+      if (!tokens[currentToken + 1].hasOwnProperty("leadingComments")) {
+        tokens[currentToken + 1].leadingComments = [];
+      }
+      tokens[currentToken + 1].leadingComments.push(element);
+    }
+  });
+
+  return tokens;
+}
+
+function findUpperBoundToken(tokens, comment) {
+  let diff;
+  let i;
+  let current;
+
+  let len = tokens.length;
+  i = 0;
+
+  while (len) {
+    diff = len >>> 1;
+    current = i + diff;
+    if (tokens[current].startOffset > comment.startOffset) {
+      len = diff;
+    } else {
+      i = current + 1;
+      len -= diff + 1;
+    }
+  }
+  return i;
+}
+
+function extendCommentRange(tokens, comments) {
+  const ignoreComments = [...comments];
+  let position;
+  ignoreComments.forEach(comment => {
+    position = findUpperBoundToken(tokens, comment);
+    comment.extendedRange = {};
+    comment.extendedRange.startOffset =
+      position - 1 < 0 ? comment.startOffset : tokens[position - 1].endOffset;
+    comment.extendedRange.endOffset =
+      position == tokens.length
+        ? comment.endOffset
+        : tokens[position].startOffset;
+  });
+  return ignoreComments;
+}
+
+function filterPrettierIgnore(comments) {
+  return [...comments].filter(comment =>
+    comment.image.match(/\/\/(\s*)prettier-ignore(\s*)/gm)
+  );
+}
+
+function shouldIgnore(node, comments, ignoredNodes) {
+  for (let i = 0; i < comments.length; i++) {
```

Comment:
are we using imperative logic for performance reasons?

Reply:
yes, as it is not necessary to continue further if the comment is already matched with the node.

Reply:
I meant why are we using for loops instead of forEach and high order functions?

Reply:
If we use a forEach it will still iterate through all elements even after the node is matched.

Reply:
Lodash's forEach can exit early: https://lodash.com/docs/4.17.11#forEach
But that is besides the point. it looks like you are trying to **find** an element in a collection and do something with that element.
- https://lodash.com/docs/4.17.11#find may be a relevant tool in this use case.

So you have to consider, do you go with the more declarative/functional style
as it is less error prone and less code, or do you go with the imperative style with regular loops and such.

Each has its pros and cons, in I personally tend to only use regular imperative loops in performance critical code, otherwise I prefer to pay a small overhead but gain higher quality source code (imho).

Reply:
Interesting, I will keep it in my mind. 👍 
I will probably do as you suggest it.
Do you have any other remark on this PR ? If not I will merge it after changing this piece of code.